app/views/course/admin/assessment_settings/edit.json.jbuilder

```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+categories = current_course.assessment_categories.includes(:tabs)
+
+json.showPublicTestCasesOutput current_course.show_public_test_cases_output
+json.showStdoutAndStderr current_course.show_stdout_and_stderr
+json.allowRandomization current_course.allow_randomization
+json.allowMrqOptionsRandomization current_course.allow_mrq_options_randomization
+
+categories_assessments_hash = Course::Assessment::Category.includes(:assessments).to_h do |category|
```

Comment:
Thanks for your feedback! I have refactored the endpoints as such:
- There are now `assessment_settings#move_assessments` and `assessment_settings#move_tabs`.
- `assessment_settings#edit` now starts from `categories`'s assessments, does 1 SQL query to get all tabs' counts, and uses calculated-attributes to generate the top 3 assessment titles.
- No more sending assessment IDs to client. Moving is handled by the new endpoints.

---

As we have discussed offline, there were two approaches for moving assessments that we talked about:
- using `update_all` (committed)
- using activerecord-import, like so:
```ruby
def move_assessments
  source_tab_id, destination_tab_id = move_assessments_params

  source_tab = Course::Assessment::Tab.find(source_tab_id)
  updated_assessments = []
  source_tab.assessments.each do |assessment|
    assessment.tab_id = destination_tab_id
    updated_assessments << assessment
  end

  ActiveRecord::Base.transaction do
    Course::Assessment.import! updated_assessments, on_duplicate_key_update: [:tab_id]
  end

  render json: { moved_assessments_count: updated_assessments.size }
rescue # whatever error here
  head :bad_request
end
```

The problem with using `update_all` is that it does not go through any ActiveModel validations or ActiveRecord updates. Moving around assessments were fine, but opening the assessment's index or edit pages now throws an error:
```
undefined method `id' for nil:NilClass
Extracted source (around line #28):

json.folder_attributes do
  json.folder_id @assessment.folder.id
  json.enable_materials_action !current_component_host[:course_materials_component].nil?
  json.materials @assessment.materials.order(:name) do |material|
    json.partial! '/course/material/material.json', material: material, folder: @assessment.folder
```
which targeted that `@assessment.folder` is `nil`. As we have discussed, this is probably caused by `update_all` skipping `Course::Assessment`'s validation that reparents the `folder` attribute `before_validation` that powers the feature that links Assessments to Materials (Workbin).

Evidently, activerecord-import triggers SQL queries for each `Course::LessonPlan::Item` to run their validations.

Let us discuss this again in due time.

Reply:
A short summary of our discussion earlier today on `move_assessments`:
- `update_all` is a really no-go because it skips the `folder` reparenting.
- activerecord-import for some reasons still does not reparent `folder`.
- For now, we just default to classic `assessment.update!` for each assessment.

We spent quite some time debugging the weird behaviour of an `assessment`'s `tab` not being updated to the destination tab despite `assessment.update!(tab_id: destination_tab_id)` in `#move_assessments`, although everything is fine when done from the Edit Assessment page.

Later I tested doing `assessment.update!(tab: destination_tab)` (using the destination `Course::Assessment::Tab` object rather than `destination_tab_id`), and **now the `folder`s reparent** ✅, according to my controller tests, that is. You may check b9162d7 for the methods, and a456f3f for the tests.

`#move_tabs` is all good ✅.